### PR TITLE
Inline styles view

### PR DESCRIPTION
**Pull request by willmcgugan**
This PR implements *inline styles*, a Styles object that overrides any rule set by the CSS and persists even if the CSS is changes.

This is accomplished by DOM nodes having two Styles objects: a base (css) Styles object, and an inline Styles object. The new RenderStyles class presents the same interface as Styles, but presents combined css+inline rules.

This PR also stores rules in the styles class differently. Previously there were private attributes for every rule. Now there is a single `_rules` dict which contains all the rules. This makes merging style objects trivial.

### CODE REVIEW

**Review by darrenburns — CHANGES_REQUESTED** [review]

File: src/textual/layouts/factory.py [hunk 1]
```diff
@@ -24,7 +25,7 @@ class MissingLayout(Exception):
     pass
 
 
-def get_layout(name: LayoutName) -> Layout:
+def get_layout(name: str) -> Layout:
```
Comment (darrenburns) [inline src/textual/layouts/factory.py:28]: How come you changed this type annotation? Also, `LayoutName` is no longer used

File: sandbox/local_styles.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        )
+
+    async def on_key(self, event: events.Key) -> None:
+        await self.dispatch_key(event)
+
+    def key_a(self) -> None:
+        self.query("#footer").set_styles(text="on magenta")
+
+    def key_b(self) -> None:
+        self["#footer"].set_styles("text: on green")
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply sandbox/local_styles.py:26]: That's a decision still to be made. Just needed something to test with.

File: src/textual/message_pump.py [hunk 3]
```diff
@@ -311,6 +311,20 @@ async def emit(self, message: Message) -> bool:
         else:
             return False
 
+    async def dispatch_key(self, event: events.Key) -> None:
```
Comment (darrenburns) [inline src/textual/message_pump.py:314]: Should a "message pump" know anything about key events? Could this method live elsewhere in the hierarchy?

File: src/textual/_timer.py [hunk 4]
```diff
@@ -5,6 +5,7 @@
     get_event_loop,
     CancelledError,
     Event,
+    shield,
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/_timer.py:8]: VScode has this annoying *feature* to automatically add imports. Even if I delete the code that references them.

File: tests/test_styles.py [hunk 5]
```diff
@@ -9,3 +10,120 @@ def test_styles_reset():
...
+    base.text_color = "yellow"
+    assert styles_view.text == Style.parse("bold red")
+
+    # Base is bold blue
+    inline.text_color = None
+    assert styles_view.text == Style.parse("bold yellow")
+
+
+def test_render_styles_border():
+
```
Comment (darrenburns) [inline tests/test_styles.py:111]:
```suggestion
```

File: tests/test_styles.py [hunk 6]
```diff
@@ -9,3 +10,120 @@ def test_styles_reset():
...
+    styles = Styles()
+    styles.set_rule("text_style", Style.parse("bold"))
+    styles.merge_rules({"display": "none"})
+    assert styles.get_rules() == {
+        "text_style": Style.parse("bold"),
+        "display": "none",
+    }
+
+
+def test_render_styles_text():
```
Comment (darrenburns) [inline tests/test_styles.py:83]: This test is great documentation :)

File: src/textual/css/stylesheet.py [hunk 7]
```diff
@@ -172,18 +173,61 @@ def apply(self, node: DOMNode) -> None:
...
-            (name, max(specificity_rules, key=get_first_item)[1])
-            for name, specificity_rules in rule_attributes.items()
-        ]
+        node_rules = cast(
+            RulesMap,
+            {
+                name: max(specificity_rules, key=get_first_item)[1]
+                for name, specificity_rules in rule_attributes.items()
+            },
+        )
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/css/stylesheet.py:182]: Mypy requires it. It's not smart enough to know that the dict contains only the allowed keys.

File: src/textual/css/styles.py [hunk 8]
```diff
@@ -131,166 +149,271 @@ class Styles:
...
-        attribute: str,
-        value: Any,
-        start_time: float,
-        duration: float | None,
-        speed: float | None,
-        easing: EasingFunction,
-    ) -> Animation | None:
-        from ..widget import Widget
+    def get_transition(self, key: str) -> Transition | None:
+        if key in self.ANIMATABLE:
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/css/styles.py:249]: Not currently... A transition could be specified on a value that is not animatable. In that case we want to return `None` and not the transition.

File: src/textual/css/styles.py [hunk 9]
```diff
@@ -131,166 +149,271 @@ class Styles:
...
+    def clear_rule(self, rule: str) -> None:
+        self._rules.pop(rule, None)
 
-    def get_transition(self, key: str) -> Transition | None:
-        if key in self.ANIMATABLE:
-            return self.transitions.get(key, None)
+    def get_rules(self) -> RulesMap:
+        return self._rules.copy()
+
+    def set_rule(self, rule: str, value: object | None) -> None:
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/css/styles.py:277]: You may be correct. Let me try that in a subsequent PR.

File: src/textual/css/styles.py [hunk 10]
```diff
@@ -131,166 +149,271 @@ class Styles:
...
+        Args:
+            other (Styles): A Styles object.
+        """
+
+        self._rules.update(other._rules)
+
+    def merge_rules(self, rules: RulesMap) -> None:
+        self._rules.update(rules)
 
     def extract_rules(
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/css/styles.py:318]: Will do. Might also make this private at some point.

File: src/textual/css/styles.py [hunk 11]
```diff
@@ -131,166 +149,271 @@ class Styles:
...
         """
-        result = cls()
-        for name in INTERNAL_RULE_NAMES:
-            setattr(result, name, getattr(style1, name) or getattr(style2, name))
-        return result
+
+        has_rule = rules.__contains__
+        get_rule = rules.__getitem__
+
+        has_top = has_rule(f"{name}_top")
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/css/styles.py:380]: True, but this isn't critical code. More for debugging currently. Will consider that for a later update.

File: src/textual/dom.py [hunk 12]
```diff
@@ -31,17 +32,20 @@ class DOMNode(MessagePump):
 
     """
 
+    DEFAULT_STYLES = ""
     STYLES = ""
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/dom.py:36]: Yeah, I reckon that's clearer.

File: src/textual/message_pump.py (hunk 3 above)
Comment (willmcgugan) [reply src/textual/message_pump.py:314]: The only only other option is DOM Node, but I'd like to keep that free of event handlers.

You're right though, it is a little incongruent. I'll leave a comment to think about that one later.

**Pull request merged**